sklearn/linear_model/_logistic.py

```diff
@@ -696,28 +316,34 @@ def _logistic_regression_path(
     # multinomial case this is not necessary.
     if multi_class == "ovr":
         w0 = np.zeros(n_features + int(fit_intercept), dtype=X.dtype)
-        mask_classes = np.array([-1, 1])
         mask = y == pos_class
         y_bin = np.ones(y.shape, dtype=X.dtype)
-        y_bin[~mask] = -1.0
-        # for compute_class_weight
+        if solver in ["lbfgs", "newton-cg"]:
+            mask_classes = np.array([0, 1])
+            y_bin[~mask] = 0.0
+        else:
+            mask_classes = np.array([-1, 1])
+            y_bin[~mask] = -1.0
 
+        # for compute_class_weight
         if class_weight == "balanced":
             class_weight_ = compute_class_weight(
                 class_weight, classes=mask_classes, y=y_bin
             )
             sample_weight *= class_weight_[le.fit_transform(y_bin)]
 
     else:
-        if solver not in ["sag", "saga"]:
+        if solver in ["sag", "saga", "lbfgs", "newton-cg"]:
+            # SAG, lbfgs and newton-cg multinomial solvers need LabelEncoder,
+            # not LabelBinarizer, i.e. y is mapped to integers.
+            le = LabelEncoder()
+            Y_multi = le.fit_transform(y).astype(X.dtype, copy=False)
```

Comment:
This saves memory compared to `LabelBinarizer`, in particular for large `n_classes`.

Reply:
Should we also add a comment in this case?
```suggestion
        if solver in ["sag", "saga", "lbfgs", "newton-cg"]:
            # SAG, lbfgs and newton-cg multinomial solvers need y represented with integers.
            # Hence we used LabelEncoder and not LabelBinarizer.
            # LabelEncoder also allows saving some memory, especially when `n_classes` is large.
            le = LabelEncoder()
            Y_multi = le.fit_transform(y).astype(X.dtype, copy=False)
```